Enhance Sunricher DALI with stale-device cleanup

niracler opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Implement automatic device cleanup when device list change


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

niracler committed:
Add options flow with gateway refresh functionality to sunricher_dali

- Add OptionsFlowHandler to allow users to refresh gateway configuration
- Support re-discovery of gateway IP address (useful for DHCP environments)
- Implement device and entity cleanup during refresh
- Add reload mechanism with configurable delays
- Include translation strings for options flow UI
- Register options flow handler in config flow class

niracler committed:
Add comprehensive options flow tests with 100% coverage

- Add 10 tests covering all options flow scenarios
- Test refresh success, errors, device cleanup, reload failures
- Add missing data_description translation
- Achieve 100% coverage for config_flow.py (120/120 lines)

niracler committed:
Refactor sunricher_dali options flow

- Replace broad Exception with specific DaliGatewayError
- Remove redundant logging and comments
- Move messages to strings.json for i18n
- Remove unused string keys
- Simplify code structure

niracler committed:
Refine sunricher DALI cleanup

niracler committed:
Fix options flow tests and add stale device removal test

- Add missing data_description translation for refresh option
- Refactor options flow tests to use async_reload mock
- Add test for device discovery failure
- Add test for stale device removal on device list decrease
- Improve test coverage to 95%

niracler committed:
Merge branch 'dev' into feat/option-flow

niracler committed:
Fix options flow leaving gateway disconnected on refresh errors

niracler committed:
Merge branch 'feat/option-flow' of https://github.com/niracler/ha-core into feat/option-flow

niracler committed:
Merge remote-tracking branch 'upstream/dev' into feat/option-flow

joostlek commented on homeassistant/components/sunricher_dali/config_flow.py:
```diff
@@ -28,12 +29,100 @@
+)
+
+
+class OptionsFlowHandler(config_entries.OptionsFlow):
```
This shouldn't be an Options flow.
1. we have a reconfigure flow to reconfigure the device. But, the device should be the same, so it should have the same serial number.
2. I don't see why people need to use this to rediscover devices? As in, when I reload the integration it would add the devices we find right? In theory it would be nice if we could add them on runtime, but we shouldn't have the user do this manually

niracler committed:
Merge remote-tracking branch 'upstream/dev' into feat/option-flow

niracler committed:
Improve Sunricher DALI gateway IP address handling

Replace manual IP refresh option with automatic rediscovery on connection failure

niracler committed:
refactor: reconnect in ip is same

niracler committed:
refactor: update testcase

niracler committed:
Merge remote-tracking branch 'upstream/dev' into feat/option-flow

niracler committed:
Merge remote-tracking branch 'upstream/dev' into feat/option-flow

niracler committed:
refactor(sunricher_dali): replace gateway discovery with DHCP-based IP lookup

- Add DHCP discovery support in config flow
- Store MAC address for IP tracking
- Use DHCP helpers instead of library discovery
- Update tests for new discovery mechanism

niracler committed:
refactor: remove unnecessary code and log

niracler committed:
refactor: remove unnecessary code and log

niracler commented:
And now, I already to use DHCP discovery.

joostlek reviewed (changes requested):
Please split the PR in the dhcp and the stale devices :)

niracler committed:
Split: drop DHCP handling

niracler commented:
> Please split the PR in the dhcp and the stale devices :)

Already split the PR

1.  stale devices: here
2. dhcp: https://github.com/home-assistant/core/pull/157798

niracler committed:
merge

joostlek reviewed (changes requested):
1. should we update the quality scale yaml file?
2. Is there a way we can find out if a light is removed on runtime?

niracler committed:
merge dev branch

joostlek committed:
Fix
